virtual_rainforest/core/grid.py

```diff
@@ -446,3 +429,129 @@ def populate_distances(
+            a given point can be empty - when the point falls in no cell - or of length
+            > 1 when a point falls on adjoining cell boundaries.
+        """
+
+        if (x_coords.ndim != 1) or (y_coords.ndim != 1):
+            raise ValueError("The x/y coordinate arrays are not 1 dimensional")
+
+        if x_coords.shape != y_coords.shape:
+            raise ValueError("The x/y coordinates are of unequal length")
+
+        # Get shapely points for the coordinates
+        xyp = [Point(x, y) for x, y in zip(x_coords, y_coords)]
+
+        # Map the Cell ID of each point - this is doing all pairs of points and cells,
+        # which is a computational choke point. Might be able to use a spatial search
+        # option, if this gets problematic, possibly including an STRTree in the grid
+        #    object https://shapely.readthedocs.io/en/latest/strtree.html
+
+        return [
+            [id for id, ply in zip(self.cell_id, self.polygons) if ply.intersects(pt)]
+            for pt in xyp
+        ]
+
+    def map_xy_to_cell_indexing(
+        self,
+        x_coords: np.ndarray,
+        y_coords: np.ndarray,
+        x_idx: Optional[np.ndarray],
+        y_idx: Optional[np.ndarray],
+        strict: bool = True,
+    ) -> tuple[np.ndarray, np.ndarray]:
+        """Returns indexing to map xy coordinates a single cell_id axis.
+
+        This function maps the provided one-dimensional set of x and y points onto the
+        grid (using `~virtual_rainforest.core.grid.Grid.map_xy_to_cell_id`) and then
+        checks that the mapped points uniquely identify all of the grid cells. If
+        `strict` is set to False, then extra points outside the grid are permitted.
+
+        The function then returns a pair of arrays that give indices on the original
+        x and y data to extract data along a single cell id axis. If specific x and y
+        indexing data is not provided, they are assumed to be simple sequences along
+        `x_coords` and `y_coords`.
+
+        Args:
+            cell_map: An existing mapping from points onto cell ids.
+            strict: Should points outside the grid be allowed
+
+        Returns:
+            A list showing the integer cell id for each pair of points. If strict is
+            False, the list element for points outside the grid is None.
+        """
+
+        # Get the coordinate mapping to cell ids
+        cell_map = self.map_xy_to_cell_id(x_coords=x_coords, y_coords=y_coords)
+
+        # Set indexing to sequence along coords if missing
+        if (x_idx is None) ^ (y_idx is None):  # Note: ^ is xor
+            raise ValueError("Only one of x/y indices provided.")
+
+        if (x_idx is None) and (y_idx is None):
+            x_idx = np.arange(x_coords.shape[0])
+            y_idx = np.arange(y_coords.shape[0])
+
+        # Check provided - note that "type: ignore" here and below comes from the
+        # Optional type of the x_idx and y_idx args. These now cannot be None.
```

Comment:
You could explicitly remove the `Optional`ness of `x_idx` and `y_idx` by using `typing.cast`.

Reply:
New to me! I've had a look and it seems there is an alternative form of using `assert`. That seems clearer to me - it seems to signal to `mypy` that from that point on that the type of the variable no longer includes None:

``` python
assert x_idx is not None
assert y_idx is not None
if (x_idx.shape != x_coords.shape) or (y_idx.shape != y_coords.shape):
    raise ValueError("Dimensions of x/y indices do not match coordinates")
```

I do think that is neater than the `# type: ignore` though.

Reply:
Actually, I've rethought this. I'm not convinced that adding  actual lines of code into the project to get `mypy` to ignore cases is better than using `type: ignore` comments to get it to shut up. 

I'm going to revert this for now but we should start a discussion to come up with a consistent approach.  @jacobcook1995 - you've used `type: ignore` - thoughts?  I note that we can set `mypy` so that `# type: ignore` _requires_ a specific problem to ignore, which might be a sensible thing to start using: e.g. ` # type: ignore [union-attr]`

Reply:
The only place I've used `type: ignore` is on package imports for packages that don't have type hints (even on `typeshed`). My concern with using `type: ignore` is whether there are downstream consequences to using it. Obviously not a problem here where a bool is just generated for a logical operation and isn't used further, but worry about not propagating type information through the module

Reply:
Can you copy that over onto #151